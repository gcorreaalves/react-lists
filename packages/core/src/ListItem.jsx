import React from 'react'
import PropTypes from 'prop-types'

const ListItem = ({ item, renderer, onClick }) => {
  const handleClick = item => e => {
    e.preventDefault()
    onClick && onClick(item)
  }

  const ItemComponent = renderer
  let props = { ...item }

  if (typeof item === 'string') {
    props = { item: item }
  }

  return (
    <li onClick={handleClick(item)}>
      <ItemComponent {...props} />
    </li>
  )
}

ListItem.propTypes = {
  item: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  renderer: PropTypes.func.isRequired,
  onClick: PropTypes.func
}

export default ListItem
