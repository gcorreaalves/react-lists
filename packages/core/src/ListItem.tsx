import React from 'react'

type Item = string | unknown

interface PropTypes {
  item: Item
  renderer: React.FunctionComponent
  onClick?: (item: Item) => void
}

const ListItem: React.FunctionComponent<PropTypes> = ({
  item,
  renderer,
  onClick
}) => {
  const handleClick = (item: Item) => (event: React.MouseEvent) => {
    event.preventDefault()
    onClick && onClick(item)
  }

  const ItemComponent = renderer
  let props

  if (typeof item === 'string') {
    props = { item: item }
  } else {
    props = { ...item }
  }

  return (
    <li onClick={handleClick(item)}>
      <ItemComponent {...props} />
    </li>
  )
}

export default ListItem
