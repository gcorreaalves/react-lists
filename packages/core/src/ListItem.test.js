import React from 'react'
import { mount, render } from 'enzyme'
import ListItem from './ListItem'

describe('ListItem component', () => {
  it('should render correctly', () => {
    const wrapper = render(
      <ListItem item='Item' renderer={({ item }) => item} />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render custom element', () => {
    const wrapper = render(
      <ListItem
        item='Item'
        renderer={({ item }) => <div className={'custom-item'}>{item}</div>}
      />
    )

    expect(wrapper.find('.custom-item').length).toBe(1)
  })

  it('should call onClick', () => {
    const handleClick = jest.fn()
    const wrapper = mount(
      <ListItem
        item='Item'
        renderer={({ item }) => item}
        onClick={handleClick}
      />
    )

    wrapper.simulate('click')

    expect(handleClick).toHaveBeenCalledWith('Item')
  })
})
