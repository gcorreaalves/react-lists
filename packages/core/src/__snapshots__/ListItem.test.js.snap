// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component should render correctly 1`] = `
<li>
  Item
</li>
`;
